tests/components/enphase_envoy/test_init.py

```diff
@@ -0,0 +1,231 @@
+from homeassistant.components.enphase_envoy.const import Platform
+from homeassistant.components.enphase_envoy.coordinator import SCAN_INTERVAL
+from homeassistant.config_entries import ConfigEntryState
+from homeassistant.const import (
+    CONF_HOST,
+    CONF_NAME,
+    CONF_PASSWORD,
+    CONF_TOKEN,
+    CONF_USERNAME,
+    STATE_UNAVAILABLE,
+)
+from homeassistant.core import HomeAssistant
+
+from . import setup_integration
+
+from tests.common import MockConfigEntry, async_fire_time_changed
+
+
+async def test_with_pre_v7_firmware(
+    hass: HomeAssistant,
+    mock_envoy: AsyncMock,
+    config_entry: MockConfigEntry,
+) -> None:
+    """Test enphase_envoy coordinator with pre V7 firmware."""
+    mock_envoy.firmware = "5.1.1"
+    mock_envoy.auth = EnvoyLegacyAuth(
+        "127.0.0.1", username="test-username", password="test-password"
+    )
+    await setup_integration(hass, config_entry)
+
+    assert config_entry.state is ConfigEntryState.LOADED
+
+    assert (entity_state := hass.states.get("sensor.inverter_1"))
+    assert entity_state.state == "1"
+
+
+@pytest.mark.freeze_time("2024-07-23 00:00:00+00:00")
+async def test_token_in_config_file(
+    hass: HomeAssistant,
+    mock_envoy: AsyncMock,
+) -> None:
+    """Test coordinator with token provided from config."""
+    token = encode(
+        payload={"name": "envoy", "exp": 1907837780},
+        key="secret",
+        algorithm="HS256",
+    )
+    entry = MockConfigEntry(
+        domain=DOMAIN,
+        entry_id="45a36e55aaddb2007c5f6602e0c38e72",
+        title="Envoy 1234",
+        unique_id="1234",
+        data={
+            CONF_HOST: "1.1.1.1",
+            CONF_NAME: "Envoy 1234",
+            CONF_USERNAME: "test-username",
+            CONF_PASSWORD: "test-password",
+            CONF_TOKEN: token,
+        },
+    )
+    mock_envoy.auth = EnvoyTokenAuth("127.0.0.1", token=token, envoy_serial="1234")
+    await setup_integration(hass, entry)
+    await hass.async_block_till_done(wait_background_tasks=True)
+    assert entry.state is ConfigEntryState.LOADED
+
+    assert (entity_state := hass.states.get("sensor.inverter_1"))
+    assert entity_state.state == "1"
+
+
+@respx.mock
+@pytest.mark.freeze_time("2024-07-23 00:00:00+00:00")
+async def test_expired_token_in_config(
+    hass: HomeAssistant,
+    mock_envoy: AsyncMock,
+) -> None:
+    """Test coordinator with expired token provided from config."""
+    new_token = encode(
+        payload={"name": "envoy", "exp": 2007837780},
+        key="secret",
+        algorithm="HS256",
+    )
+    # mock token authorization from pyenphase to test token refresh code
+    with patch(
+        "pyenphase.auth.EnvoyTokenAuth._obtain_token",
+        return_value=new_token,
+    ):
+        current_token = encode(
+            # some time in 2021
+            payload={"name": "envoy", "exp": 1627314600},
+            key="secret",
+            algorithm="HS256",
+        )
```

Comment:
What kind of effect does this have on the library and the returned objects?

Reply:
This replaces the method in the library that will connect to the Enphase cloud and retrieve a token to use for authentication in the local Envoy device. Previous version patched the httpx request this method was doing. This version still returns a token as if it came from the Enphase Cloud.

```py
    async def _obtain_token(self) -> str:
        """Obtain the token for Envoy authentication."""
.
.
            response = await self._post_json_with_cloud_client(
.
.
            return response.text
```